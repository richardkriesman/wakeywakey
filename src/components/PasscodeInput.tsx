/**
 * @module components
 */

import { AntDesign } from "@expo/vector-icons";
import React, { ReactNode } from "react";
import ReactNative, { StyleSheet, View } from "react-native";
import { Input, Text } from "react-native-elements";

const PASSCODE_LENGTH = 4;

export interface PasscodeInputProps {
    // optional
    confirmPasscode: boolean; // whether to confirm, does nothing if mustEqual is set
    defaultConfirmText: string;
    defaultPromptText: string;
    mustEqual: string; // passcode must equal this for success

    // required
    handleSuccess(passcode: string): void;
}

export interface PasscodeInputState {
    errorMessage: string;
    isConfirming: boolean;
    passcode: string;
    promptText: string;
}

/**
 * PasscodeInput
 * This component allows the user to enter a passcode.
 * You can set whether to confirm or match password to a set password.
 * When success, handleSuccess is called.
 */

export class PasscodeInput extends React.Component<PasscodeInputProps, PasscodeInputState> {
    public static defaultProps = {
        confirmPasscode: false,
        defaultConfirmText: "Confirm your passcode:",
        defaultPromptText: "Enter passcode:",
        mustEqual: ""
    };

    private inputRef = React.createRef<Input>();

    public constructor(props: PasscodeInputProps) {
        super(props);
    }

    public componentWillMount(): void {
        this.setState({
            errorMessage: "",
            isConfirming: false,
            passcode: "",
            promptText: this.props.defaultPromptText
        });
    }

    public handleChangeText(): void {
        if (this.state.errorMessage.length !== 0) {
            this.setState({ errorMessage: "" });
        }
    }

    public handleSubmitEditing(event:
            ReactNative.NativeSyntheticEvent<ReactNative.TextInputSubmitEditingEventData>): void {
        const passcode: string = event.nativeEvent.text;

        if (passcode.length !== PASSCODE_LENGTH) { // passcode too short
            this.setState({
                errorMessage: "Passcode must be " + PASSCODE_LENGTH + " digits long!"
            });
        } else { // passcode long enough

            // check state of input (choosing or confirming)
            if (!this.state.isConfirming) {
                this.handleChoosingPasscode(passcode);
            } else {
                this.handleConfirmingPasscode(passcode);
            }
        }
    }

    /**
     * handles choosing passcode
     * Sets the state to confirming.
     * If mustEqual is set, checks if passcode is equal. If not equal, resets.
     */
    public handleChoosingPasscode(passcode: string): void {
        if (this.props.mustEqual !== "") {
            if (passcode === this.props.mustEqual) { // correct passcode
                this.props.handleSuccess(passcode); // done!
            } else { // incorrect passcode
                this.setState({
                    errorMessage: "Incorrect passcode!",
                    passcode: ""
                });
                this.inputRef.current.clear();
            }
        } else {
            if (this.props.confirmPasscode) { // switch to confirming
                this.setState({
                    isConfirming: true,
                    passcode,
                    promptText: this.props.defaultConfirmText
                });
                this.inputRef.current.clear();
            } else { // not confirming passcode
                this.props.handleSuccess(passcode);
            }
        }
    }

    /**
     * handles confirming passcode
     * Checks if passcode is correct and calls handleSuccess if yes.
     */
    public handleConfirmingPasscode(passcode: string): void {
        if (passcode === this.state.passcode) { // correct passcode
            this.props.handleSuccess(passcode); // done!
        } else { // incorrect passcode
            this.setState({
                errorMessage: "Passcodes don't match!",
                isConfirming: false,
                passcode: "",
                promptText: this.props.defaultPromptText
            });
            this.inputRef.current.clear();
        }
    }

    public render(): ReactNode {
        return (
            <View style={styles.mainContainer}>
                <Text style={styles.promptStyle}>
                    {this.state.promptText}
                </Text>
                <Input
                    containerStyle={styles.containerStyle}
                    inputContainerStyle={styles.inputContainerStyle}
                    inputStyle={styles.inputStyle}

                    autoFocus={true}
                    blurOnSubmit={false}
                    caretHidden={true}
                    errorMessage={this.state.errorMessage}
                    keyboardType="numeric"
                    leftIcon={ <AntDesign name="lock" size={32} /> }
                    maxLength={PASSCODE_LENGTH}
                    onChangeText={this.handleChangeText.bind(this)}
                    onSubmitEditing={this.handleSubmitEditing.bind(this)}
                    ref={this.inputRef}
                    secureTextEntry={true}
                />
            </View>
        );
    }

}

const styles = StyleSheet.create({
    containerStyle: {
        backgroundColor: "white",
        borderRadius: 5,
        marginTop: 20,
        width: "70%"
    },
    errorStyle: {
        color: "red",
        textAlign: "center"
    },
    inputContainerStyle: {
        marginBottom: 5,
        marginTop: 5
    },
    inputStyle: {
        fontSize: 32,
        letterSpacing: 30,
        marginLeft: 10,
        textAlign: "left"
    },
    mainContainer: {
        alignItems: "center",
        bottom: 50,
        flex: 1,
        justifyContent: "center",
        width: "100%"
    },
    promptStyle: {
        color: "black",
        fontSize: 32,
        fontWeight: "bold",
        textAlign: "center"
    }
});
