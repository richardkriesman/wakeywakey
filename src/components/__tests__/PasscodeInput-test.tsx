import * as React from "react";
import renderer from "react-test-renderer";
import { PasscodeInput } from "../PasscodeInput";

describe("PasscodeInput", () => {

    let props: any;
    let component: any;

    beforeEach(() => {
        props = { handleSuccess: jest.fn() };
        component = renderer.create(<PasscodeInput {...props} />).getInstance();
        component.setState({ passcode: "1234" });
    });

    it("renders correctly", () => {
        const tree = renderer.create(<PasscodeInput {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("handles changeText with short passcode", () => {
        component.setState({ errorMessage: "..." });
        const setState = jest.spyOn(component, "setState");
        component.handleChangeText("123");
        expect(setState).toBeCalled();
    });

    it("handles changeText with correct length passcode", () => {
        const choosingPasscode = jest.spyOn(component, "handleChoosingPasscode");
        const confirmingPasscode = jest.spyOn(component, "handleConfirmingPasscode");
        component.handleChangeText("1234");
        expect(choosingPasscode).toBeCalled();
        component.setState({ isConfirming: true });
        component.handleChangeText("1234");
        expect(confirmingPasscode).toBeCalled();
    });

    it("handles submit editing", () => {
        const setState = jest.spyOn(component, "setState");
        component.handleSubmitEditing({ nativeEvent: { text: "1234" } });
        expect(setState).not.toBeCalled();
        component.handleSubmitEditing({ nativeEvent: { text: "123" } });
        expect(setState).toBeCalled();
    });

    it("choosing passcode with verify", async (done) => {
        props = {
            handleSuccess: jest.fn(),
            verifyPasscode: jest.fn().mockImplementation(async (passcode) => passcode === "1234")
        };
        component = renderer.create(<PasscodeInput {...props} />).getInstance();
        const setState = jest.spyOn(component, "setState");
        await component.handleChoosingPasscode("1234"); // correct passcode
        expect(props.handleSuccess).toBeCalled();
        await component.handleChoosingPasscode("5555"); // incorrect passcode
        expect(setState).toBeCalled();
        expect(props.verifyPasscode).toBeCalled();

        done();
    });

    it("choosing passcode without confirm", () => {
        component.handleChoosingPasscode("1234");
        expect(props.handleSuccess).toBeCalled();
    });

    it("choosing passcode with confirm", () => {
        props = { confirmPasscode: true, handleSuccess: jest.fn() };
        component = renderer.create(<PasscodeInput {...props} />).getInstance();
        const setState = jest.spyOn(component, "setState");
        component.handleChoosingPasscode("1234");
        expect(setState).toBeCalled();
    });

    it("confirming correct passcode", () => {
        component.handleConfirmingPasscode("1234");
        expect(props.handleSuccess).toBeCalled();
    });

    it("confirming incorrect passcode", () => {
        const setState = jest.spyOn(component, "setState");
        component.handleConfirmingPasscode("5555");
        expect(setState).toBeCalled();
    });
});
