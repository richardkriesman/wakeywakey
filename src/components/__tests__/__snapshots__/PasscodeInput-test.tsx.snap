// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasscodeInput renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "flex-end",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ffffff",
        "fontSize": 32,
        "paddingBottom": 20,
        "paddingTop": 40,
        "textAlign": "center",
      }
    }
    theme={
      Object {
        "colors": Object {
          "disabled": "hsl(208, 8%, 90%)",
          "divider": "#bcbbc1",
          "error": "#ff190c",
          "grey0": "#393e42",
          "grey1": "#43484d",
          "grey2": "#5e6977",
          "grey3": "#86939e",
          "grey4": "#bdc6cf",
          "grey5": "#e1e8ee",
          "greyOutline": "#bbb",
          "platform": Object {
            "android": Object {
              "error": "#f44336",
              "primary": "#2196f3",
              "secondary": "#9C27B0",
              "success": "#4caf50",
              "warning": "#ffeb3b",
            },
            "ios": Object {
              "error": "#ff3b30",
              "primary": "#007aff",
              "secondary": "#5856d6",
              "success": "#4cd964",
              "warning": "#ffcc00",
            },
          },
          "primary": "#2089dc",
          "searchBg": "#303337",
          "secondary": "#8F0CE8",
          "success": "#52c41a",
          "warning": "#faad14",
        },
      }
    }
  >
    Enter passcode:
  </Text>
  <View
    style={
      Object {
        "paddingHorizontal": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
          "width": "70%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 40,
            "justifyContent": "center",
            "marginLeft": 15,
          }
        }
      >
        <Text />
      </View>
      <TextInput
        allowFontScaling={true}
        autoFocus={true}
        blurOnSubmit={false}
        caretHidden={true}
        contextMenuHidden={true}
        keyboardType="numeric"
        maxLength={4}
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        returnKeyType="done"
        secureTextEntry={true}
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "flex": 1,
            "fontSize": 30,
            "letterSpacing": 20,
            "marginRight": 32,
            "minHeight": 40,
            "textAlign": "center",
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;
